examples/feature_demo/video_yuv.py

```diff
@@ -0,0 +1,267 @@
+    FILENAME = imageio.core.Request("imageio:cockatoo_yuv420p.mp4", "r?").filename
+
+
+def video_width_height():
+    container = av.open(FILENAME)
+    return (container.streams[0].width, container.streams[0].height)
+
+
+def benchmark_video_read():
+    start_time = time.perf_counter()
+    with av.open(FILENAME) as container:
+        frames_read = 0
+        for frame in container.decode(video=0):
+            frame = frame.reformat(format=FORMAT)
+            frame.to_ndarray()
+            frames_read += 1
+
+            if frames_read > 100:
+                break
+    end_time = time.perf_counter()
+    elapsed_time = end_time - start_time
+    frames_per_second = frames_read / elapsed_time
+    return frames_per_second
+
+
+if "PYTEST_CURRENT_TEST" not in os.environ:
+    # A mini benchmark to show the limits of just reading in the video from storage
+    for _ in range(5):
+        print(
+            f"Reading video in {FORMAT} format: {benchmark_video_read():.2f} frames per second"
+        )
+
+
+def loop_video():
+    while True:
+        with av.open(FILENAME) as container:
+            for frame in container.decode(video=0):
+                # Reformat if necessary. If format is the same, reformat is a no-op.
+                # Otherwise, this includes a data copy and some computations.
+                # In the case of yuv420 -> rgb24, the image also takes more memory.
+                frame = frame.reformat(format=FORMAT)
+                # Cast to numpy array. This should not involve a data copy.
+                yield frame.to_ndarray()
+
+
+w, h = video_width_height()
+frame_generator = loop_video()
+
+# Create image object
+
+if FORMAT == "yuv420p":
+    # For planar yuv420, we use a 2-layer grayscale texture. The u and v planes
+    # fit in a single layer since they're half the size (in both dimensions).
+    tex = gfx.Texture(
+        size=(w, h, 2),
+        dim=2,
+        colorspace="yuv420p",
+        colorrange=COLORRANGE,
+        format="r8unorm",
+        usage=wgpu.TextureUsage.COPY_DST,
+    )
+elif FORMAT == "yuv444p":
+    tex = gfx.Texture(
+        size=(w, h, 3),
+        dim=2,
+        colorspace="yuv444p",
+        colorrange=COLORRANGE,
+        format="r8unorm",
+        usage=wgpu.TextureUsage.COPY_DST,
+    )
+else:
+    # For rgb/rgba we use an rgba texture (there is no rgb texture format).
+    tex = gfx.Texture(
+        size=(w, h, 1),
+        dim=2,
+        colorspace="srgb",
+        format="rgba8unorm",
+        usage=wgpu.TextureUsage.COPY_DST,
+    )
+
+im = gfx.Image(
+    gfx.Geometry(grid=tex),
+    gfx.ImageBasicMaterial(clim=(0, 255)),
+)
+
+# Setup the rest of the viz
+
+CanvasClass = OffscreenCanvas if OFFSCREEN else WgpuCanvas
+canvas = CanvasClass(size=(w // 2, h // 2), max_fps=999, vsync=False)
+renderer = gfx.renderers.WgpuRenderer(canvas, show_fps=True)
+scene = gfx.Scene()
+scene.add(im)
+camera = gfx.OrthographicCamera(w, h)
+camera.local.position = w // 2, h // 2, 0
```

Comment:
I take it the idea here was to ensure that we render every. single. pixel. of. the. video.

How can I make sure the internal texture is the right size?

For example, your M1 uses fractional scaling, and so does my modern laptop.

Reply:
That was not even intentional. The default canvas size 640x480, I changed that 640x360 to give it the same aspect ratio as the video. The fact that, with a pixel_ratio of 2, this is exactly the same as the video resolution was a coincidence 😅 

BTW on a screen that is not hidpi, the internal texture will be twice the size, so the "every-single-pixel-is-rendered applies to any screen (except perhaps when someone has a weird OS pixel_ratio). In any case, I don't think it matters much; the rendering is not the bottleneck.

Reply:
ok, well i made it so.